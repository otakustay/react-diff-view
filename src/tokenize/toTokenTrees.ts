import {flatMap, keyBy} from 'lodash';
import type {AST, RefractorNode, highlight} from 'refractor';
import {computeOldLineNumber, computeNewLineNumber, ChangeData, HunkData} from '../utils';
import {Pair, TokenNode} from './interface';

interface Refactor {
    highlight: typeof highlight;
}

interface HunkApplyState {
    newStart: number;
    changes: ChangeData[];
}

// This function mutates `linesOfCode` argument.
function applyHunk(linesOfCode: string[], {newStart, changes}: HunkApplyState) {
    // Within each hunk, changes are continous, so we can use a sequential algorithm here.
    //
    // When `linesOfCode` is received here, it has already patched by previous hunk,
    // thus the starting line number has changed due to possible unbanlanced deletions and insertions,
    // we should use `newStart` as the first line number of current reduce.
    const [patchedLines] = changes.reduce<[string[], number]>(
        ([lines, cursor], {content, isInsert, isDelete}) => {
            if (isDelete) {
                lines.splice(cursor, 1);
                return [lines, cursor];
            }

            if (isInsert) {
                lines.splice(cursor, 0, content);
            }
            return [lines, cursor + 1];
        },
        [linesOfCode, newStart - 1]
    );

    return patchedLines;
}

function applyDiff(oldSource: string, hunks: HunkData[]): string {
    // `hunks` must be ordered here.
    const patchedLines = hunks.reduce(applyHunk, oldSource.split('\n'));
    return patchedLines.join('\n');
}

function mapChanges<T>(changes: ChangeData[], side: 'old' | 'new', toValue: (change: ChangeData | undefined) => T): T[] {
    if (!changes.length) {
        return [];
    }

    const computeLineNumber = side === 'old' ? computeOldLineNumber : computeNewLineNumber;
    const changesByLineNumber = keyBy(changes, computeLineNumber);
    const maxLineNumber = computeLineNumber(changes[changes.length - 1]);
    // TODO: why don't we start from the first change's line number?
    return Array.from({length: maxLineNumber}).map((value, i) => toValue(changesByLineNumber[i + 1]));
}

function groupChanges(hunks: HunkData[]): Pair<ChangeData[]> {
    const changes = flatMap(hunks, hunk => hunk.changes);
    return changes.reduce<[ChangeData[], ChangeData[]]>(
        ([oldChanges, newChanges], change) => {
            if (change.isNormal) {
                oldChanges.push(change);
                newChanges.push(change);
            }
            else if (change.isDelete) {
                oldChanges.push(change);
            }
            else {
                newChanges.push(change);
            }

            return [oldChanges, newChanges];
        },
        [[], []]
    );
};

function toTextPair(hunks: HunkData[]): Pair<string> {
    const [oldChanges, newChanges] = groupChanges(hunks);
    const toText = (change: ChangeData | undefined) => (change ? change.content : '');
    const oldText = mapChanges(oldChanges, 'old', toText).join('\n');
    const newText = mapChanges(newChanges, 'new', toText).join('\n');
    return [oldText, newText];
}

function createRoot(children: RefractorNode[]): TokenNode {
    return {type: 'root', children: children};
}

export interface ToTokenTreeOptionsNoHighlight {
    highlight?: false;
    oldSource?: string;
}

export interface ToTokenTreeOptionsHighlight {
    highlight: true;
    refractor: Refactor;
    oldSource?: string;
    language: string;
}

export type ToTokenTreeOptions = ToTokenTreeOptionsNoHighlight | ToTokenTreeOptionsHighlight;

export default function toTokenTrees(hunks: HunkData[], options: ToTokenTreeOptions): Pair<TokenNode> {
    if (options.oldSource) {
        const newSource = applyDiff(options.oldSource, hunks);
        const highlightText = options.highlight
            ? (text: string) => options.refractor.highlight(text, options.language)
            : (text: string): AST.Text[] => [{type: 'text', value: text}];

        return [
            createRoot(highlightText(options.oldSource)),
            createRoot(highlightText(newSource)),
        ];
    }

    const [oldText, newText] = toTextPair(hunks);
    const toTree = options.highlight
        ? (text: string) => createRoot(options.refractor.highlight(text, options.language))
        : (text: string) => createRoot([{type: 'text', value: text}]);

    return [toTree(oldText), toTree(newText)];
}
