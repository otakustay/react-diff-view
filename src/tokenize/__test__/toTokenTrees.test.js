import refractor from 'refractor';
import toTokenTrees from '../toTokenTrees';

// eslint-disable-next-line max-len
const content = '<img style="width:<?php echo (80 / count($images)) ?>%" src=\'<?php echo Component_Union_FinanceInfoModel::UNION_LICEPIC_URL_PREFIX . $image?>\'/>';

describe('toTokenTrees', () => {
    test('php will renders __PHP__', () => {
        const options = {
            highlight: true,
            refractor: refractor,
            language: 'php',
            oldSource: null,
            enhancers: [],
        };

        const hunks = [{
            changes: [
                {
                    content,
                    isInsert: true,
                    lineNumber: 1,
                    type: 'insert',
                },
            ],
            content: '@@ -1,10 +1,17 @@',
            isPlain: false,
            newLines: 17,
            newStart: 1,
            oldLines: 10,
            oldStart: 1,
        }];

        const tokens = toTokenTrees(hunks, options);
        expect(tokens).toMatchSnapshot();
    });

    test('refractor highlight', () => {
        expect(refractor.highlight(content, 'php')).toMatchSnapshot();
    });
});
