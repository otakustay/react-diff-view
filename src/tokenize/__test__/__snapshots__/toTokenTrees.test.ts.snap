// Vitest Snapshot v1

exports[`toTokenTrees > php will renders __PHP__ 1`] = `
[
  {
    "children": [],
    "type": "root",
  },
  {
    "children": [
      {
        "children": [
          {
            "children": [
              {
                "children": [
                  {
                    "type": "text",
                    "value": "<",
                  },
                ],
                "properties": {
                  "className": [
                    "token",
                    "punctuation",
                  ],
                },
                "tagName": "span",
                "type": "element",
              },
              {
                "type": "text",
                "value": "img",
              },
            ],
            "properties": {
              "className": [
                "token",
                "tag",
              ],
            },
            "tagName": "span",
            "type": "element",
          },
          {
            "children": [
              {
                "children": [
                  {
                    "type": "text",
                    "value": " ",
                  },
                  {
                    "children": [
                      {
                        "type": "text",
                        "value": "style",
                      },
                    ],
                    "properties": {
                      "className": [
                        "token",
                        "attr-name",
                      ],
                    },
                    "tagName": "span",
                    "type": "element",
                  },
                ],
                "properties": {
                  "className": [
                    "token",
                    "attr-name",
                  ],
                },
                "tagName": "span",
                "type": "element",
              },
              {
                "children": [
                  {
                    "type": "text",
                    "value": "=\\"",
                  },
                ],
                "properties": {
                  "className": [
                    "token",
                    "punctuation",
                  ],
                },
                "tagName": "span",
                "type": "element",
              },
              {
                "children": [
                  {
                    "children": [
                      {
                        "type": "text",
                        "value": "width",
                      },
                    ],
                    "properties": {
                      "className": [
                        "token",
                        "property",
                      ],
                    },
                    "tagName": "span",
                    "type": "element",
                  },
                  {
                    "children": [
                      {
                        "type": "text",
                        "value": ":",
                      },
                    ],
                    "properties": {
                      "className": [
                        "token",
                        "punctuation",
                      ],
                    },
                    "tagName": "span",
                    "type": "element",
                  },
                  {
                    "type": "text",
                    "value": "___PHP",
                  },
                  {
                    "children": [
                      {
                        "type": "text",
                        "value": "0",
                      },
                    ],
                    "properties": {
                      "className": [
                        "token",
                        "number",
                      ],
                    },
                    "tagName": "span",
                    "type": "element",
                  },
                  {
                    "type": "text",
                    "value": "___%",
                  },
                ],
                "properties": {
                  "className": [
                    "token",
                    "attr-value",
                  ],
                },
                "tagName": "span",
                "type": "element",
              },
              {
                "children": [
                  {
                    "type": "text",
                    "value": "\\"",
                  },
                ],
                "properties": {
                  "className": [
                    "token",
                    "punctuation",
                  ],
                },
                "tagName": "span",
                "type": "element",
              },
            ],
            "properties": {
              "className": [
                "token",
                "style-attr",
                "language-css",
              ],
            },
            "tagName": "span",
            "type": "element",
          },
          {
            "type": "text",
            "value": " ",
          },
          {
            "children": [
              {
                "type": "text",
                "value": "src",
              },
            ],
            "properties": {
              "className": [
                "token",
                "attr-name",
              ],
            },
            "tagName": "span",
            "type": "element",
          },
          {
            "children": [
              {
                "children": [
                  {
                    "type": "text",
                    "value": "=",
                  },
                ],
                "properties": {
                  "className": [
                    "token",
                    "punctuation",
                  ],
                },
                "tagName": "span",
                "type": "element",
              },
              {
                "children": [
                  {
                    "type": "text",
                    "value": "'",
                  },
                ],
                "properties": {
                  "className": [
                    "token",
                    "punctuation",
                  ],
                },
                "tagName": "span",
                "type": "element",
              },
              {
                "type": "text",
                "value": "___PHP1___",
              },
              {
                "children": [
                  {
                    "type": "text",
                    "value": "'",
                  },
                ],
                "properties": {
                  "className": [
                    "token",
                    "punctuation",
                  ],
                },
                "tagName": "span",
                "type": "element",
              },
            ],
            "properties": {
              "className": [
                "token",
                "attr-value",
              ],
            },
            "tagName": "span",
            "type": "element",
          },
          {
            "children": [
              {
                "type": "text",
                "value": "/>",
              },
            ],
            "properties": {
              "className": [
                "token",
                "punctuation",
              ],
            },
            "tagName": "span",
            "type": "element",
          },
        ],
        "properties": {
          "className": [
            "token",
            "tag",
          ],
        },
        "tagName": "span",
        "type": "element",
      },
    ],
    "type": "root",
  },
]
`;

exports[`toTokenTrees > refractor highlight 1`] = `
[
  {
    "children": [
      {
        "children": [
          {
            "children": [
              {
                "type": "text",
                "value": "<",
              },
            ],
            "properties": {
              "className": [
                "token",
                "punctuation",
              ],
            },
            "tagName": "span",
            "type": "element",
          },
          {
            "type": "text",
            "value": "img",
          },
        ],
        "properties": {
          "className": [
            "token",
            "tag",
          ],
        },
        "tagName": "span",
        "type": "element",
      },
      {
        "children": [
          {
            "children": [
              {
                "type": "text",
                "value": " ",
              },
              {
                "children": [
                  {
                    "type": "text",
                    "value": "style",
                  },
                ],
                "properties": {
                  "className": [
                    "token",
                    "attr-name",
                  ],
                },
                "tagName": "span",
                "type": "element",
              },
            ],
            "properties": {
              "className": [
                "token",
                "attr-name",
              ],
            },
            "tagName": "span",
            "type": "element",
          },
          {
            "children": [
              {
                "type": "text",
                "value": "=\\"",
              },
            ],
            "properties": {
              "className": [
                "token",
                "punctuation",
              ],
            },
            "tagName": "span",
            "type": "element",
          },
          {
            "children": [
              {
                "children": [
                  {
                    "type": "text",
                    "value": "width",
                  },
                ],
                "properties": {
                  "className": [
                    "token",
                    "property",
                  ],
                },
                "tagName": "span",
                "type": "element",
              },
              {
                "children": [
                  {
                    "type": "text",
                    "value": ":",
                  },
                ],
                "properties": {
                  "className": [
                    "token",
                    "punctuation",
                  ],
                },
                "tagName": "span",
                "type": "element",
              },
              {
                "type": "text",
                "value": "___PHP",
              },
              {
                "children": [
                  {
                    "type": "text",
                    "value": "0",
                  },
                ],
                "properties": {
                  "className": [
                    "token",
                    "number",
                  ],
                },
                "tagName": "span",
                "type": "element",
              },
              {
                "type": "text",
                "value": "___%",
              },
            ],
            "properties": {
              "className": [
                "token",
                "attr-value",
              ],
            },
            "tagName": "span",
            "type": "element",
          },
          {
            "children": [
              {
                "type": "text",
                "value": "\\"",
              },
            ],
            "properties": {
              "className": [
                "token",
                "punctuation",
              ],
            },
            "tagName": "span",
            "type": "element",
          },
        ],
        "properties": {
          "className": [
            "token",
            "style-attr",
            "language-css",
          ],
        },
        "tagName": "span",
        "type": "element",
      },
      {
        "type": "text",
        "value": " ",
      },
      {
        "children": [
          {
            "type": "text",
            "value": "src",
          },
        ],
        "properties": {
          "className": [
            "token",
            "attr-name",
          ],
        },
        "tagName": "span",
        "type": "element",
      },
      {
        "children": [
          {
            "children": [
              {
                "type": "text",
                "value": "=",
              },
            ],
            "properties": {
              "className": [
                "token",
                "punctuation",
              ],
            },
            "tagName": "span",
            "type": "element",
          },
          {
            "children": [
              {
                "type": "text",
                "value": "'",
              },
            ],
            "properties": {
              "className": [
                "token",
                "punctuation",
              ],
            },
            "tagName": "span",
            "type": "element",
          },
          {
            "type": "text",
            "value": "___PHP1___",
          },
          {
            "children": [
              {
                "type": "text",
                "value": "'",
              },
            ],
            "properties": {
              "className": [
                "token",
                "punctuation",
              ],
            },
            "tagName": "span",
            "type": "element",
          },
        ],
        "properties": {
          "className": [
            "token",
            "attr-value",
          ],
        },
        "tagName": "span",
        "type": "element",
      },
      {
        "children": [
          {
            "type": "text",
            "value": "/>",
          },
        ],
        "properties": {
          "className": [
            "token",
            "punctuation",
          ],
        },
        "tagName": "span",
        "type": "element",
      },
    ],
    "properties": {
      "className": [
        "token",
        "tag",
      ],
    },
    "tagName": "span",
    "type": "element",
  },
]
`;
